### PR TITLE
[CUMULUS-4089]: Scaffold Duplicate Granule Integration Tests

### PR DESCRIPTION
**Pull request by npauzenga**
**Summary:** Summary of changes

Addresses CUMULUS-4089 https://bugs.earthdata.nasa.gov/browse/CUMULUS-4089

What it does:
- Adds new integration test files for DiscoverGranules and Ingest with Unique Granule IDs
- Sets the above test files up with data and ensure the basic structure passes CI
- Adds stubbed assertions based on https://wiki.earthdata.nasa.gov/display/CUMULUS/Duplicate+Granules%3A+Testable+Requirements and related docs
- Modify existing integration test files for Lzards, Orca, CNM
- Modify existing Ingest and CNM workflows to include CHOICE block for uniquification, driven by meta config
- Modify DiscoverGranules workflow to pass `uniquifyGranuleId` and related variables. DG not yet updated to use them.

What it doesn't do:
- Make actual assertions and test cases work
- Provide a super rigid structure for tests. The stubbed assertions may need to be ordered differently based on the test logic once implemented.

### CODE REVIEW

File: tf-modules/cumulus/outputs.tf [hunk 1]
```diff
@@ -51,7 +51,7 @@ output "add_missing_file_checksums_task" {
   value = module.ingest.add_missing_file_checksums_task
 }
 
-output "add_unique_granule_id_arn" {
+output "add_unique_granule_id_task" {
```
Comment (npauzenga) [inline tf-modules/cumulus/outputs.tf:54]: Updating to be consistent with the other task outputs.

File: example/spec/parallel/ingestGranule/IngestGranuleWithDuplicateHandlingUniqueSpec.js [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+    ]);
+  });
+
+  beforeEach(() => {
+    if (beforeAllError) fail(beforeAllError);
+  });
+
+  it('prepares the test suite successfully', () => {
+    failOnSetupError([beforeAllError]);
+  });
```
Comment (npauzenga) [inline example/spec/parallel/ingestGranule/IngestGranuleWithDuplicateHandlingUniqueSpec.js:252]: We're about here. The test sets up, the step function runs, the new `addUniqueGranuleId` task is executed but it's returning with an empty payload ` "payload": {},`. The input contains the granules as expected but they don't make it to the output. The task's logs show that the output contains the granules so it seems like something is wrong between the return of `assignUniqueIds` and the output of the task.

File: example/spec/parallel/cnmWorkflow/CnmWorkflowFromSqsSpec.js [hunk 3]
```diff
@@ -276,4 +276,27 @@ describe('The Cloud Notification Mechanism SQS workflow', () => {
       await expectAsync(waitForQueueMessageCount(queues.sourceQueueUrl, 0)).toBeResolved();
     });
   });
+
+  describe('a duplicate granule record', () => {
+    beforeAll(() => {
+      if (beforeAllFailed) return;
+      try {
+        ruleOverride.meta.uniquifyGranuleId = true;
```
Comment (npauzenga) [inline example/spec/parallel/cnmWorkflow/CnmWorkflowFromSqsSpec.js]: This doesn't (and likely shouldn't) work. I will find a different way to update the rule.

File: example/cumulus-tf/cnm_workflow.asl.json [hunk 4]
```diff
@@ -50,6 +50,43 @@
           "Next": "CnmResponseFail"
         }
       ],
+      "Next": "ChooseUniqueGranules"
+    },
+    "ChooseUniqueGranules": {
```
Comment (Nnaga1) [inline example/cumulus-tf/cnm_workflow.asl.json]: question for my own curiousity, are ChooseUniqueGranules and AddUniqueGranuleId separate tasks?

File: example/cumulus-tf/discover_granules_workflow.asl.json [hunk 5]
```diff
@@ -16,7 +16,9 @@
             "collection": "{$.meta.collection}",
             "buckets": "{$.meta.buckets}",
             "stack": "{$.meta.stack}",
-            "duplicateGranuleHandling": "{$.meta.collection.duplicateHandling}"
+            "duplicateGranuleHandling": "{$.meta.collection.duplicateHandling}",
+            "uniquifyGranuleId": "{$.meta.collection.meta.uniquifyGranuleId}",
+            "hashLength": "{$.meta.collection.meta.hashLength}"
```
Comment (Nnaga1) [inline example/cumulus-tf/discover_granules_workflow.asl.json:21]: what does hashLength do? for the uniquification hash used?

File: example/cumulus-tf/discover_granules_workflow.asl.json [hunk 6]
```diff
@@ -16,7 +16,9 @@
...
           }
```
Comment (Nnaga1) [inline example/cumulus-tf/discover_granules_workflow.asl.json:22]: does this not need to be added to: https://github.com/nasa/cumulus/blob/eeb799f8b13c618d8a5a50fed650b8d17f2d8ba3/example/cumulus-tf/discover_granules_workflow.tf#L4 here?

File: example/spec/parallel/cnmWorkflow/CnmWorkflowFromSqsSpec.js [hunk 7]
```diff
@@ -276,4 +276,27 @@ describe('The Cloud Notification Mechanism SQS workflow', () => {
...
+        console.log('beforeAll error', error);
+        beforeAllFailed = error;
+      }
+    });
+
+    it('executes successfully', () => {
+      if (beforeAllFailed) fail(beforeAllFailed);
+      expect(executionStatus).toEqual('SUCCEEDED');
+    });
+    xit('is successfully ingested', () => {
```
Comment (Nnaga1) [inline example/spec/parallel/cnmWorkflow/CnmWorkflowFromSqsSpec.js:296]: its just scaffolding for now right? so its not part of the ticket to actually test it (bc the uniqify task doesnt exist?)

File: example/spec/parallel/cnmWorkflow/data/rules/sqs/ASCATB_L2_Coastal_sqsRule.json [hunk 8]
```diff
@@ -8,7 +8,8 @@
   "meta": {
     "cnmResponseStream": "{{cnmResponseStream}}",
     "cnmResponseMethod": "sns",
-    "cmrMetadataFormat": "umm_json_v1_6_2"
+    "cmrMetadataFormat": "umm_json_v1_6_2",
+    "uniquifyGranuleId": false
```
Comment (Nnaga1) [inline example/spec/parallel/cnmWorkflow/data/rules/sqs/ASCATB_L2_Coastal_sqsRule.json]: this should be false for this:`"uniquifyGranuleId"`, for this specific workflow or just in general?

File: example/spec/parallel/discoverGranules/DiscoverGranulesDuplicateHandlingUniqueSpec.js [hunk 9]
```diff
@@ -0,0 +1,171 @@
+'use strict';
+
+const {
+  loadCollection,
+  loadProvider,
+} = require('@cumulus/integration-tests');
+const { LambdaStep } = require('@cumulus/integration-tests/sfnStep');
+const {
+  createCollection, deleteCollection,
```
Comment (Nnaga1) [inline example/spec/parallel/discoverGranules/DiscoverGranulesDuplicateHandlingUniqueSpec.js]: I would reccomend to space the imports on different lines, like you did in line 4-5, just a NIT

File: example/spec/parallel/discoverGranules/DiscoverGranulesDuplicateHandlingUniqueSpec.js [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+        filename: './data/collections/s3_MOD09GQ_006/s3_MOD09GQ_006.json',
+        postfix: testId,
+      });
+
+      await createCollection({ prefix: stackName, collection });
+
+      providerPath = `cumulus-test-data/${testId}`;
+
+      // Upload the granule to be discovered
+      await uploadS3GranuleDataForDiscovery({
```
Comment (Nnaga1) [inline example/spec/parallel/discoverGranules/DiscoverGranulesDuplicateHandlingUniqueSpec.js:76]: should there not be multiple granules?

File: example/spec/parallel/discoverGranules/DiscoverGranulesDuplicateHandlingUniqueSpec.js [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+      });
+
+      // Execute the DiscoverGranules workflow
+      workflowExecution = await buildAndExecuteWorkflow(
+        stackName,
+        bucket,
+        'DiscoverGranules',
+        collection,
+        provider,
+        undefined,
```
Comment (Nnaga1) [inline example/spec/parallel/discoverGranules/DiscoverGranulesDuplicateHandlingUniqueSpec.js:88]: can you do like variable: undefined, for the paramlist here, if that works, just so we know what it is, or would that be unneccessary?

File: example/spec/parallel/discoverGranules/DiscoverGranulesDuplicateHandlingUniqueSpec.js [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+      if (beforeAllError) fail(beforeAllError);
+    });
+    it('executes successfully', async () => {});
+    it('results in a successful IngestGranule workflow execution', async () => {});
+    it('it makes the granule available via the Cumulus API', async () => {});
+    it('it publishes the granule metadata to CMR', async () => {});
+  });
+
+  // TODO: do we need all of these assertions for the second workflow execution?
+  //  Do we want to just include the duplicate granule in the above workflow execution?
```
Comment (Nnaga1) [inline example/spec/parallel/discoverGranules/DiscoverGranulesDuplicateHandlingUniqueSpec.js:163]: a bit confused at the question, my bad, for this:
`Do we want to just include the duplicate granule in the above workflow execution?` do you mean include it **at all** or just include it **here or elsewhere**? I would agree more assertions the better but if its doing the same thing maybe not?

File: example/spec/parallel/orca/OrcaBackupAndRecoverySpec.js [hunk 13]
```diff
@@ -149,6 +149,11 @@ describe('The S3 Ingest Granules workflow', () => {
     });
   });
 
+  describe('Uniquified duplicate granules', () => {
```
Comment (Nnaga1) [inline example/spec/parallel/orca/OrcaBackupAndRecoverySpec.js:152]: 👍

File: example/spec/parallel/ingestGranule/IngestGranuleWithDuplicateHandlingUniqueSpec.js [hunk 14]
```diff
@@ -0,0 +1,357 @@
+'use strict';
```
Comment (Nnaga1) [inline example/spec/parallel/ingestGranule/IngestGranuleWithDuplicateHandlingUniqueSpec.js:1]: the name of the test: IngestGranuleWithDuplicateHandlingUniqueSpec.js why not `IngestGranuleWithDuplicateHandlingSpec` , just a bit but made it a bit hard to understand

File: example/spec/parallel/ingestGranule/IngestGranuleWithDuplicateHandlingUniqueSpec.js [hunk 15]
```diff
@@ -0,0 +1,357 @@
...
+  const collectionDupeHandling = 'error';
+
+  let beforeAllError;
+  let collection;
+  let collectionId;
+  let config;
+  let expectedPayload;
+  let expectedS3TagSet;
+  let expectedSyncGranulePayload;
+  // let granuleCompletedMessageKey;
```
Comment (Nnaga1) [inline example/spec/parallel/ingestGranule/IngestGranuleWithDuplicateHandlingUniqueSpec.js:70]: is this the WIP thing remaining you mentioned?

File: example/spec/parallel/ingestGranule/IngestGranuleWithDuplicateHandlingUniqueSpec.js [hunk 16]
```diff
@@ -0,0 +1,357 @@
...
+        s3().putObjectTagging({ Bucket: config.bucket, Key: `${fileToTag.path}/${fileToTag.name}`, Tagging: { TagSet: expectedS3TagSet } })));
+
+      const templatedSyncGranuleFilename = templateFile({
+        inputTemplateFilename: './spec/parallel/ingestGranule/SyncGranule.output.payload.template.json',
+        config: {
+          granules: [
+            {
+              files: [
+                {
+                  bucket: config.buckets.internal.name,
```
Comment (Nnaga1) [inline example/spec/parallel/ingestGranule/IngestGranuleWithDuplicateHandlingUniqueSpec.js:124]: all the files in the same bucket? thats intentional right?

File: example/spec/parallel/ingestGranule/IngestGranuleWithDuplicateHandlingUniqueSpec.js [hunk 17]
```diff
@@ -0,0 +1,357 @@
...
+        }
+      );
+    } catch (error) {
+      beforeAllError = error;
+    }
+  });
+
+  afterAll(async () => {
+    // granule may already have been deleted by
+    // granule deletion spec. but in case that spec
```
Comment (Nnaga1) [inline example/spec/parallel/ingestGranule/IngestGranuleWithDuplicateHandlingUniqueSpec.js:205]: 👍

File: example/spec/parallel/ingestGranule/IngestGranuleWithDuplicateHandlingUniqueSpec.js (hunk 2 above)
Comment (Nnaga1) [reply example/spec/parallel/ingestGranule/IngestGranuleWithDuplicateHandlingUniqueSpec.js:252]: thats expected right? `addUniqueGranuleId` doesnt exist yet this just sets up testing for that

File: example/cumulus-tf/cnm_workflow.asl.json (hunk 4 above)
Comment (npauzenga) [reply example/cumulus-tf/cnm_workflow.asl.json]: Well `ChooseUniqueGranules` is better described as a 'state'. It represents branching logic for the Step Function to switch between tasks depending on a variable but isn't actually a task itself. `AddUniqueGranuleId` is a full-fledged task with code.

File: example/cumulus-tf/discover_granules_workflow.asl.json (hunk 5 above)
Comment (npauzenga) [reply example/cumulus-tf/discover_granules_workflow.asl.json:21]: Yup! That's a new parameter for the uniquification task. In this context (DiscoverGranules), that logic hasn't been implemented so it does... nothing... but that's the intention.

File: example/cumulus-tf/discover_granules_workflow.asl.json (hunk 6 above)
Comment (npauzenga) [reply example/cumulus-tf/discover_granules_workflow.asl.json:22]: I don't think so. These are _collection_ configuration options that will be passed via the collection's meta. We _shouldn't_ need terraform variables unless we end up adding a task to the workflow.

...is what I'm asserting. There aren't tests yet for functionality here that isn't implement yet sooooooooooo......

File: example/spec/parallel/cnmWorkflow/CnmWorkflowFromSqsSpec.js (hunk 7 above)
Comment (npauzenga) [reply example/spec/parallel/cnmWorkflow/CnmWorkflowFromSqsSpec.js:296]: Right. Should be this: https://bugs.earthdata.nasa.gov/browse/CUMULUS-4119

The task _does_ exist so this one and the Ingest spec could be started whenever.

File: example/spec/parallel/cnmWorkflow/data/rules/sqs/ASCATB_L2_Coastal_sqsRule.json (hunk 8 above)
Comment (npauzenga) [reply example/spec/parallel/cnmWorkflow/data/rules/sqs/ASCATB_L2_Coastal_sqsRule.json]: The default, insofar as this is the previous behavior, should be `false`. I'm adding it here to this rule mostly just to be explicit.

File: example/spec/parallel/discoverGranules/DiscoverGranulesDuplicateHandlingUniqueSpec.js (hunk 9 above)
Comment (npauzenga) [reply example/spec/parallel/discoverGranules/DiscoverGranulesDuplicateHandlingUniqueSpec.js]: https://github.com/nasa/cumulus/pull/3985/commits/34cdc75318464986df300c7168c21f5f1aac53fd

File: example/spec/parallel/discoverGranules/DiscoverGranulesDuplicateHandlingUniqueSpec.js (hunk 10 above)
Comment (npauzenga) [reply example/spec/parallel/discoverGranules/DiscoverGranulesDuplicateHandlingUniqueSpec.js:76]: Yes but I haven't added all of the granules that will probably be required. We can discuss if that's in scope here. Because I don't totally know the shape of the DG work, I didn't want to add too much prematurely but that could be the wrong approach.

I did update the comment and skip those assertions that aren't asserting. https://github.com/nasa/cumulus/pull/3985/commits/9ffaf8e59a6f8e8d30a834fa8c6958cdc49244c8

File: example/spec/parallel/ingestGranule/IngestGranuleWithDuplicateHandlingUniqueSpec.js (hunk 14 above)
Comment (npauzenga) [reply example/spec/parallel/ingestGranule/IngestGranuleWithDuplicateHandlingUniqueSpec.js:1]: Certainly could but this type of "uniquification" handling is just one option. I was following the naming convention of the already-created specs:

- /ingestGranule/IngestGranuleWithDuplicateHandlingErrorSpec.js
- /ingestGranule/IngestGranuleWithDuplicateHandlingVersionSpec.js

File: example/spec/parallel/ingestGranule/IngestGranuleWithDuplicateHandlingUniqueSpec.js (hunk 15 above)
Comment (npauzenga) [reply example/spec/parallel/ingestGranule/IngestGranuleWithDuplicateHandlingUniqueSpec.js:70]: Yeah, it's related to https://github.com/nasa/cumulus/pull/3985/files/fad56ceb2a4f7dbcb1dc15b5ee8d6a114e0196f0#diff-7cf08dcd03b23ecb59c1e30d23ddd41ad65ad3458b2eccfe44699c1e994f1dc9R241

File: example/spec/parallel/ingestGranule/IngestGranuleWithDuplicateHandlingUniqueSpec.js (hunk 16 above)
Comment (npauzenga) [reply example/spec/parallel/ingestGranule/IngestGranuleWithDuplicateHandlingUniqueSpec.js:124]: Yeah, this matches the behavior in other ingest specs 🤷

**Comment by npauzenga** [comment]
Thanks @Nnaga1 

>thats expected right? addUniqueGranuleId doesnt exist yet this just sets up testing for that

This spec should actually pass. The task does exist and is used in the new Ingest workflow so the suite should set up and tear down.

File: example/cumulus-tf/cnm_workflow.asl.json (hunk 4 above)
Comment (Nnaga1) [reply example/cumulus-tf/cnm_workflow.asl.json]: ohhh so its like a boolean: ChooseUniqueGranules, false do one thing, true do another, in the way of workflow inputs/outputs, that makes sense 🙌

File: example/cumulus-tf/discover_granules_workflow.asl.json (hunk 6 above)
Comment (Nnaga1) [reply example/cumulus-tf/discover_granules_workflow.asl.json:22]: and I guess if this isnt the case: `These are collection configuration options that will be passed via the collection's meta. ` (which seems like not a high percentage maybe?) then the terraform can be added if needed? but yea your assertion makes sense 👍 I think its fine to leave it alone

File: example/spec/parallel/cnmWorkflow/data/rules/sqs/ASCATB_L2_Coastal_sqsRule.json (hunk 8 above)
Comment (Nnaga1) [reply example/spec/parallel/cnmWorkflow/data/rules/sqs/ASCATB_L2_Coastal_sqsRule.json]: so if you removed this: `"uniquifyGranuleId": false` it would work as expected? if so yea I agree its good to be explicit, just asking (no change :) )

File: example/spec/parallel/discoverGranules/DiscoverGranulesDuplicateHandlingUniqueSpec.js (hunk 10 above)
Comment (Nnaga1) [reply example/spec/parallel/discoverGranules/DiscoverGranulesDuplicateHandlingUniqueSpec.js:76]: thats fair, the amount of granules used I guess can easily be increased, so if the 4119 needs more than one they can add it but if its good with one then its fine, I guess its not really on you to decide 1 or 2 or 3 or .... concretely, just set it up, makes sense 🙌

File: example/spec/parallel/ingestGranule/IngestGranuleWithDuplicateHandlingUniqueSpec.js (hunk 14 above)
Comment (Nnaga1) [reply example/spec/parallel/ingestGranule/IngestGranuleWithDuplicateHandlingUniqueSpec.js:1]: maybe `UniqifySpec` 🤣 jk, I guess its fine, it follows the convention so, nvm 👍

File: example/cumulus-tf/cnm_workflow.asl.json [hunk 18]
```diff
@@ -50,6 +50,43 @@
           "Next": "CnmResponseFail"
         }
       ],
+      "Next": "ChooseUniqueGranules"
```
Comment (npauzenga) [inline example/cumulus-tf/cnm_workflow.asl.json]: Of note here that there's another PR with these changes https://github.com/nasa/cumulus/pull/3999/files#diff-9256c5750029f6d4bbf1b585ea8a5cfeb475cd18c74b975d7cbfbee840f3b481

I cleaned this one up a bit to hopefully conflict less. The changes in the above linked PR are more relevant as it contains the logic to make it work. This PR should support integration testing the CNM workflow but a lot of that setup is done.

Comment (Nnaga1) [reply example/cumulus-tf/cnm_workflow.asl.json]: and It would be finished in the 4119 ticket or following ticket which deals with that (since this is only scaffolding and the task hasnt been implemented yet), sounds good 👍

File: CHANGELOG.md [hunk 19]
```diff
@@ -12,7 +12,8 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
   - Updated the `parse-pdr` task component to throw an error if multiple granules within the same PDR have the same granuleId after applying the granuleIdFilter, unless the `uniquifyGranuleId` configuration parameter is explicitly set to `true`.
 
 ### Added
-
+- **CUMULUS-4089**
+  - Add integration testing for duplicate granule workflows
```
Comment (Jkovarik) [inline CHANGELOG.md]: Nit: can we be specific re: which tests?

File: example/cumulus-tf/cnm_workflow.asl.json [hunk 20]
```diff
@@ -50,6 +50,43 @@
...
+    },
+    "ChooseUniqueGranules": {
+      "Choices": [
+        {
+          "Next": "AddUniqueGranuleId",
+          "BooleanEquals": true,
+          "Variable": "$.meta.uniquifyGranuleId"
```
Comment (Jkovarik) [inline example/cumulus-tf/cnm_workflow.asl.json]: Nit: Could we namespace this?  I realize we're just throwing all kinds of other things in .meta, but it's never really been a good thing for the base message.

File: example/cumulus-tf/cnm_workflow.asl.json (hunk 4 above)
Comment (npauzenga) [reply example/cumulus-tf/cnm_workflow.asl.json]: Yeah, it's a "state" in the state machine that looks for a boolean to make a decision.

File: example/cumulus-tf/discover_granules_workflow.asl.json (hunk 6 above)
Comment (npauzenga) [reply example/cumulus-tf/discover_granules_workflow.asl.json:22]: Yeah, well, even in the case where _everyone_ wants this, it still probably wouldn't be a tf change. They'd have to configure their collections with the appropriate meta variable, not the workflow. If I'm understanding your question correctly.

Comment (Nnaga1) [reply example/cumulus-tf/discover_granules_workflow.asl.json:22]: it does, meta of the collection msg config not TF, thanks 👍

**Comment by Jkovarik** [comment]
General comment: if there are tests where the entire suite is `xit`, should be be running the before/setup at all until we *do* have assertions being tested?

File: example/spec/parallel/discoverGranules/DiscoverGranulesDuplicateHandlingUniqueSpec.js (hunk 11 above)
Comment (npauzenga) [reply example/spec/parallel/discoverGranules/DiscoverGranulesDuplicateHandlingUniqueSpec.js:88]: I think parameterizing this function would be a good idea. But it'd also be a change that updates all of the integration specs if we change the function's signature. The changeset would be an additional 25-30 files. If you'd like to argue that that's appropriate here I definitely can 😬

File: example/spec/parallel/discoverGranules/DiscoverGranulesDuplicateHandlingUniqueSpec.js (hunk 12 above)
Comment (npauzenga) [reply example/spec/parallel/discoverGranules/DiscoverGranulesDuplicateHandlingUniqueSpec.js:163]: This is kind of a question for the person that implements the DG logic. The options here as I see it are:
- Re-run the worfklow with duplicates
- Include duplicates in the workflow execution at the top of this test suite

Both have pros and cons and I have placed these stubs in a way that implies the second approach but didn't want to be super prescriptive as it doesn't _really_ matter.

File: CHANGELOG.md (hunk 19 above)
Comment (npauzenga) [reply CHANGELOG.md]: Yes. https://github.com/nasa/cumulus/pull/3985/commits/82ad420d39459343b3d3ed599bbe19942cb37ca6

File: example/cumulus-tf/cnm_workflow.asl.json (hunk 20 above)
Comment (npauzenga) [reply example/cumulus-tf/cnm_workflow.asl.json]: I should back these changes out in favor of https://github.com/nasa/cumulus/pull/3999/files#diff-d7e739d439b6ee346b45ccb58afa057bb4ec3f2ed9ed9c7f4fab9f124eff2944

Comment (npauzenga) [reply example/cumulus-tf/cnm_workflow.asl.json]: Done: https://github.com/nasa/cumulus/pull/3985/commits/3a02d5202d407e5fd266cf17c51b4536a8ec7671

Comment (npauzenga) [reply example/cumulus-tf/cnm_workflow.asl.json]: I moved this param to the collection meta in the other specs. CI validating that I didn't mess it up now: https://github.com/nasa/cumulus/pull/3985/commits/8612a162c7d23fab99f534c6ec6821654d481f48

File: example/spec/parallel/discoverGranules/DiscoverGranulesDuplicateHandlingUniqueSpec.js (hunk 11 above)
Comment (Nnaga1) [reply example/spec/parallel/discoverGranules/DiscoverGranulesDuplicateHandlingUniqueSpec.js:88]: oh yikes yea, nvm, I guess if it follows the format/style of the others its fine 👍

File: example/spec/parallel/discoverGranules/DiscoverGranulesDuplicateHandlingUniqueSpec.js (hunk 12 above)
Comment (Nnaga1) [reply example/spec/parallel/discoverGranules/DiscoverGranulesDuplicateHandlingUniqueSpec.js:163]: so that comment was for the future person? ok

File: example/cumulus-tf/cnm_workflow.asl.json (hunk 20 above)
Comment (npauzenga) [reply example/cumulus-tf/cnm_workflow.asl.json]: Also https://github.com/nasa/cumulus/pull/3985/commits/2f38c9d1bf0a35f77846cff7de45116a37fee7f9 😬

**Comment by npauzenga** [comment]
> General comment: if there are tests where the entire suite is `xit`, should be be running the before/setup at all until we _do_ have assertions being tested?

We definitely don't _have_ to. I know this will slow down the tests a little but we will eventually make use of all of that setup before merging this to main. So the "slower for no reason" portion of this *should* be short-lived. Unless we're worried about problems with cleanup or something? But that should be largely resolved now that ES isn't a problem...?

File: example/spec/parallel/ingestGranule/IngestGranuleSuccessSpec.js [hunk 21]
```diff
@@ -1408,7 +1408,7 @@ describe('The S3 Ingest Granules workflow', () => {
         expect(definition.Comment).toEqual('Ingest Granule');
 
         // definition has all the states' information
-        expect(Object.keys(definition.States).length).toBe(12);
+        expect(Object.keys(definition.States).length).toBe(14);
```
Comment (npauzenga) [inline example/spec/parallel/ingestGranule/IngestGranuleSuccessSpec.js:1411]: This is an update because we're adding additional states for the conditional block.

File: example/cumulus-tf/cnm_workflow.asl.json [hunk 22]
```diff
@@ -50,43 +50,6 @@
           "Next": "CnmResponseFail"
         }
       ],
-      "Next": "ChooseUniqueGranules"
-    },
-    "ChooseUniqueGranules": {
```
Comment (Nnaga1) [inline example/cumulus-tf/cnm_workflow.asl.json]: so the CNM workflow doesnt need this anymore? is there a reason why or

Comment (Nnaga1) [reply example/cumulus-tf/cnm_workflow.asl.json]: oh didnt see commit msg, its another PR 👀

File: example/spec/parallel/ingestGranule/IngestGranuleSuccessSpec.js (hunk 21 above)
Comment (npauzenga) [reply example/spec/parallel/ingestGranule/IngestGranuleSuccessSpec.js:1411]: Further, it's an update because previously I had the Ingest workflow for unique granules completely separate. Now it's the same workflow so will require updates to this test.

File: example/cumulus-tf/ingest_and_publish_granule_workflow.asl.json [hunk 23]
```diff
@@ -5,9 +5,17 @@
     "ChooseUniqueGranules": {
       "Choices": [
         {
-          "Next": "AddUniqueGranuleId",
-          "BooleanEquals": true,
-          "Variable": "{$.meta.collection.meta.uniquifyGranuleId}"
+          "And": [
```
Comment (Nnaga1) [inline example/cumulus-tf/ingest_and_publish_granule_workflow.asl.json:8]: what does this "And" thing do? just for my understanding

File: example/cumulus-tf/cnm_workflow.asl.json (hunk 22 above)
Comment (npauzenga) [reply example/cumulus-tf/cnm_workflow.asl.json]: Yeah, I backed out my changes in favor of [the PR that actually implements it](https://github.com/nasa/cumulus/pull/3999)

File: example/cumulus-tf/ingest_and_publish_granule_workflow.asl.json (hunk 23 above)
Comment (npauzenga) [reply example/cumulus-tf/ingest_and_publish_granule_workflow.asl.json:8]: It's a safer configuration (again, stolen from our teammate 😆). It's saying "The variable is present AND it's true" as opposed to just "It's true" without a check for its existence.

Comment (Nnaga1) [reply example/cumulus-tf/ingest_and_publish_granule_workflow.asl.json:8]: ohhhhhhhh ok, it bundles the two together, thats kinda cool

**Comment by Jkovarik** [comment]
@npauzenga  thanks for responding to my comments - no objections here 👍🏻
